src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+use quick_xml::{
+    events::{attributes::Attributes, Event},
+    reader::Reader,
+};
+use std::{
+    borrow::Cow,
+    collections::BTreeMap,
+    io::{self, BufRead},
+    sync::Arc,
+};
+
+use super::{
+    common::{fix_partially_demangled_rust_symbol, Occurrences},
+    Collapse,
+};
+
+const REF: &[u8] = b"ref";
+const ID: &[u8] = b"id";
+const NAME: &[u8] = b"name";
+const TRACE_QUERY_RESULT: &[u8] = b"trace-query-result";
+const NODE: &[u8] = b"node";
+const ROW: &[u8] = b"row";
+const BACKTRACE: &[u8] = b"backtrace";
+const FRAME: &[u8] = b"frame";
+
+/// Context of collapsing a xctrace's `Time Profiler` xml
+#[derive(Default)]
+struct CollapseContext {
+    /// xml tag backtrace
+    state_backtrace: TagBacktrace,
+    // --------- per-xml caches below -----------
+    /// backtrace_id <--> BackTrace
+    backtraces: BTreeMap<u64, Arc<Backtrace>>,
+    /// backtrace_id <--> Frame
+    frames: BTreeMap<u64, Arc<Frame>>,
+}
+
+// Note that sometimes same Backtrace have different id, we need to merge them
+// before writing folded file.
+struct BacktraceOccurrences {
+    /// How many times the backtrace occurred.
+    num: u64,
+    /// Backtrace content
+    backtrace: Arc<Backtrace>,
+}
+
+/// Backtrace of scanning xml tags.
+#[derive(Default)]
+struct TagBacktrace {
+    backtrace: Vec<CurrentTag>,
+}
+
+impl TagBacktrace {
+    fn push_back(&mut self, tag: CurrentTag) {
+        self.backtrace.push(tag);
+    }
+
+    /// Remove the top tag and return it. Return `Err(())` on tag mismatching.
+    fn collide(&mut self, name: &[u8]) -> Result<CurrentTag, ()> {
```

Comment:
The name and description of this don't quite match. `collide` doesn't imply `pop`, and it's not clear what mismatching/colliding "means". Can I suggest: `pop_with_name`? And then have it return an `Option` instead.